internal/controller/clusterdeployment_controller_test.go

```diff
@@ -164,65 +236,357 @@ var _ = Describe("ClusterDeployment Controller", func() {
+						LastTransitionTime: metav1.Now(),
+					},
+					{
+						Type:               clusterapiv1beta1.InfrastructureReadyCondition,
+						Status:             corev1.ConditionTrue,
+						LastTransitionTime: metav1.Now(),
+					},
+				})
+				Expect(k8sClient.Status().Update(ctx, &cluster)).To(Succeed())
+
+				Expect(Get(&machineDeployment)()).To(Succeed())
+				machineDeployment.SetConditions([]clusterapiv1beta1.Condition{
+					{
+						Type:               clusterapiv1beta1.MachineDeploymentAvailableCondition,
+						Status:             corev1.ConditionTrue,
+						LastTransitionTime: metav1.Now(),
+					},
+				})
+				Expect(k8sClient.Status().Update(ctx, &machineDeployment)).To(Succeed())
 			})
-			Expect(err).NotTo(HaveOccurred())
+
+			By("ensuring ClusterDeployment is reconciled", func() {
+				Eventually(func(g Gomega) {
+					_, err := controllerReconciler.Reconcile(ctx, reconcile.Request{
+						NamespacedName: client.ObjectKeyFromObject(&clusterDeployment),
+					})
+					g.Expect(err).NotTo(HaveOccurred())
+					g.Expect(Object(&clusterDeployment)()).Should(SatisfyAll(
+						HaveField("Status.Conditions", ContainElements(
+							SatisfyAll(
+								HaveField("Type", hmc.HelmReleaseReadyCondition),
+								HaveField("Status", metav1.ConditionTrue),
+								HaveField("Reason", hcv2.InstallSucceededReason),
+							),
+							SatisfyAll(
+								HaveField("Type", hmc.SveltosProfileReadyCondition),
+								HaveField("Status", metav1.ConditionTrue),
+								HaveField("Reason", hmc.SucceededReason),
+							),
+							SatisfyAll(
+								HaveField("Type", string(clusterapiv1beta1.ControlPlaneInitializedCondition)),
+								HaveField("Status", metav1.ConditionTrue),
+								HaveField("Reason", hmc.SucceededReason),
+							),
+							SatisfyAll(
+								HaveField("Type", string(clusterapiv1beta1.ControlPlaneReadyCondition)),
+								HaveField("Status", metav1.ConditionTrue),
+								HaveField("Reason", hmc.SucceededReason),
+							),
+							SatisfyAll(
+								HaveField("Type", string(clusterapiv1beta1.InfrastructureReadyCondition)),
+								HaveField("Status", metav1.ConditionTrue),
+								HaveField("Reason", hmc.SucceededReason),
+							),
+							SatisfyAll(
+								HaveField("Type", string(clusterapiv1beta1.MachineDeploymentAvailableCondition)),
+								HaveField("Status", metav1.ConditionTrue),
+								HaveField("Reason", hmc.SucceededReason),
+							),
+							// SatisfyAll(
+							// 	HaveField("Type", hmc.FetchServicesStatusSuccessCondition),
+							// 	HaveField("Status", metav1.ConditionTrue),
+							// 	HaveField("Reason", hmc.SucceededReason),
+							// ),
+							// SatisfyAll(
+							// 	HaveField("Type", hmc.ReadyCondition),
+							// 	HaveField("Status", metav1.ConditionTrue),
+							// 	HaveField("Reason", hmc.SucceededReason),
+							// ),
+						))))
+				}).Should(Succeed())
+			})
+		})
+
+		// TODO (brongineer): Add test for ClusterDeployment reconciliation with Azure credentials
```

Comment:
what do we expect here, file dedicated issues or refactor the code to maximize its reusability and then "copy-paste" the tests? i personally think that both of the proposals are time-wise costly and in general do not have much business value (having good coverage for the tier-1 provider)

Reply:
This and following todo's needed to not to forget that atm only one branch in creds propagation is tested. And actually it tests nothing:

```go
// clusterdeployment_controller.go L715
case "aws":
    l.Info("Skipping creds propagation for AWS")
```

> what do we expect here, file dedicated issues or refactor the code to maximize its reusability and then "copy-paste" the tests? i personally think that both of the proposals are time-wise costly and in general do not have much business value (having good coverage for the tier-1 provider)

hm, can't agree. without good coverage we can't predict the behaviour. Like did you know that clusterdeployment controller panics if for some reason helmchart status does not contain artifact field?